### ISSUE TITLE
eval EXPR doesn't keep $1 afterwards

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#11242]
Migrated from [rt.perl.org#88014](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=88014) (status was 'resolved')

Searchable as RT88014$

**Comment by p5pRT** [comment]
### From @epa

#### Created by @epa



If you do a regexp match inside an eval EXPR then the capture
variables $1 etc\. are not set afterwards\.  This is even though
perlfunc says that&#8203;:

&nbsp;    In the first form\, the return value of EXPR is parsed and executed
&nbsp;    as if it were a little Perl program\.  The value of the expression
&nbsp;    \(which is itself determined within scalar context\) is first parsed\,
&nbsp;    and if there weren't any errors\, executed in the lexical context of
&nbsp;    the current Perl program\, so that any variable settings or
&nbsp;    subroutine and format definitions remain afterwards\.

Here is a test program&#8203;:

&nbsp;    \#\!/usr/bin/perl
&nbsp;    use warnings;
&nbsp;    use strict;
&nbsp;    use 5\.10\.0;

&nbsp;    $\_ = 'a';

&nbsp;    eval '/\(a\)/';
&nbsp;    my $got = $1;
&nbsp;    say $got // 'undef';

&nbsp;    /\(a\)/;
&nbsp;    $got = $1;
&nbsp;    say $got // 'undef';

I expected both to print 'a' but the first one prints 'undef'\.

\(Obviously\, for eval BLOCK the value of $1 should not be preserved\,
since it is dynamically scoped to a BLOCK\.  But eval EXPR is supposed
to execute within the current BLOCK\.\)

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

This perlbug was built using Perl 5.12.3 in the Fedora build system.
It is being executed now by Perl 5.12.3 - Thu Mar 10 18:59:35 UTC 2011.

Site configuration information for perl 5.12.3:

Configured by Red Hat, Inc. at Thu Mar 10 18:59:35 UTC 2011.

Summary of my perl5 (revision 5 version 12 subversion 3) configuration:
   
  Platform:
    osname=linux, osvers=2.6.32-71.14.1.el6.x86_64,
archname=x86_64-linux-thread-multi
    uname='linux x86-09.phx2.fedoraproject.org 2.6.32-71.14.1.el6.x86_64
#1 smp wed jan 5 17:01:01 est 2011 x86_64 x86_64 x86_64 gnulinux '
    config_args='-des -Doptimize=-O2 -g -pipe -Wall
-Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector
--param=ssp-buffer-size=4 -m64 -mtune=generic
-Dccdlflags=-Wl,--enable-new-dtags -DDEBUGGING=-g -Dversion=5.12.3
-Dmyhostname=localhost -Dperladmin=root@localhost -Dcc=gcc -Dcf_by=Red
Hat, Inc. -Dprefix=/usr -Dvendorprefix=/usr -Dsiteprefix=/usr/local
-Dsitelib=/usr/local/share/perl5 -Dsitearch=/usr/local/lib64/perl5
-Dprivlib=/usr/share/perl5 -Dvendorlib=/usr/share/perl5
-Darchlib=/usr/lib64/perl5 -Dvendorarch=/usr/lib64/perl5
-Darchname=x86_64-linux-thread-multi -Dlibpth=/usr/local/lib64 /lib64
/usr/lib64 -Duseshrplib -Dusethreads -Duseithreads
-Dusedtrace=/usr/bin/dtrace -Duselargefiles -Dd_semctl_semun -Di_db
-Ui_ndbm -Di_gdbm -Di_shadow -Di_syslog -Dman3ext=3pm -Duseperlio
-Dinstallusrbinperl=n -Ubincompat5005 -Uversiononly
-Dpager=/usr/bin/less -isr -Dd_gethostent_r_proto -Ud_endhostent_r_proto
-Ud_sethostent_r_proto -Ud_endprotoent_r_proto -Ud_setprotoent_r_proto
-Ud_endservent_r_proto -Ud_setservent_r_proto -Dscriptdir=/usr/bin
-Dotherlibdirs=/usr/local/lib64/perl5/site_perl/5.10.0/x86_64-linux-thre
ad-multi:/usr/local/lib/perl5/site_perl/5.10.0:/usr/lib64/perl5/vendor_p
erl/5.10.0/x86_64-linux-thread-multi:/usr/lib/perl5/vendor_perl:/usr/lib
/perl5/site_perl'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='gcc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing
-pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE
-D_FILE_OFFSET_BITS=64',
    optimize='-O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions
-fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -pipe
-fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.5.1 20100924 (Red Hat 4.5.1-4)',
gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t',
lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags =' -fstack-protector'
    libpth=/usr/local/lib64 /lib64 /usr/lib64
    libs=-lresolv -lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lpthread
-lc -lgdbm_compat
    perllibs=-lresolv -lnsl -ldl -lm -lcrypt -lutil -lpthread -lc
    libc=, so=so, useshrplib=true, libperl=libperl.so
    gnulibc_version='2.13'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef,
ccdlflags='-Wl,--enable-new-dtags -Wl,-rpath,/usr/lib64/perl5/CORE'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -g -pipe -Wall
-Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector
--param=ssp-buffer-size=4 -m64 -mtune=generic'

Locally applied patches:
    


@INC for perl 5.12.3:
    /home/eda/lib/perl5/5.10.0
    /home/eda/lib/perl5/site_perl/5.10.0
    /usr/local/lib64/perl5
    /usr/local/share/perl5
    /usr/lib64/perl5
    /usr/share/perl5
    /usr/lib64/perl5
    /usr/share/perl5
    /usr/local/lib64/perl5/site_perl/5.10.0/x86_64-linux-thread-multi
    /usr/local/lib/perl5/site_perl/5.10.0
    /usr/lib64/perl5/vendor_perl/5.10.0/x86_64-linux-thread-multi
    /usr/lib/perl5/vendor_perl
    /usr/lib/perl5/site_perl
    .


Environment for perl 5.12.3:
    HOME=/home/eda
    LANG=en_GB.UTF-8
    LANGUAGE (unset)
    LC_COLLATE=C
    LC_CTYPE=en_GB.UTF-8
    LC_MESSAGES=en_GB.UTF-8
    LC_MONETARY=en_GB.UTF-8
    LC_NUMERIC=en_GB.UTF-8
    LC_TIME=en_GB.UTF-8
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
 
PATH=/home/eda/bin:/home/eda/bin:/usr/lib64/qt-3.3/bin:/usr/lib64/ccache
:/usr/local/bin:/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/sbin:/sbin:/usr
/sbin:/home/eda/bin:/sbin:/usr/sbin
 
PERL5LIB=/home/eda/lib/perl5/5.10.0:/home/eda/lib/perl5/site_perl/5.10.0
    PERL_BADLANG (unset)
    SHELL=/bin/bash



______________________________________________________________________
This email has been scanned by the MessageLabs Email Security System.
For more information please visit http://www.messagelabs.com/email 
______________________________________________________________________

```
</details>

**Comment by p5pRT** [comment]
### From zefram@fysh.org

Ed Avis wrote&#8203;:
>\(Obviously\, for eval BLOCK the value of $1 should not be preserved\,
>since it is dynamically scoped to a BLOCK\.  But eval EXPR is supposed
>to execute within the current BLOCK\.\)

That's a doc bug\.  String eval does establish its own lexical and dynamic
context\, just like a normal code block\.  The documentation referring to
the lexical context in which the eval appears is trying to explain that
the lexical context inside the eval is \*nested within\* the context of
the eval\.

\-zefram

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @epa

I must admit I don't understand\.  Do you mean that there is no
difference in scoping behaviour between eval "\.\.\." and eval \{ \.\.\. \}?  A
quick test seems to suggest that is the case&#8203;:

our $x = 1;
eval "local $x = 2;";
say $x;
eval \{ local $x = 3; \};
say $x;

**Comment by p5pRT** [comment]
### From @iabyn

On Fri\, Apr 08\, 2011 at 10&#8203;:38&#8203;:28AM \+0100\, Zefram wrote&#8203;:
> Ed Avis wrote&#8203;:
> >\(Obviously\, for eval BLOCK the value of $1 should not be preserved\,
> >since it is dynamically scoped to a BLOCK\.  But eval EXPR is supposed
> >to execute within the current BLOCK\.\)
> 
> That's a doc bug\.  String eval does establish its own lexical and dynamic
> context\, just like a normal code block\.  The documentation referring to
> the lexical context in which the eval appears is trying to explain that
> the lexical context inside the eval is \*nested within\* the context of
> the eval\.


I've fixed perlfunc with commit 2341804c04aefb306adfdbe22c23e6cd11f0e48f\.

\-\- 
"There's something wrong with our bloody ships today\, Chatfield\."
&nbsp;    \-\- Admiral Beatty at the Battle of Jutland\, 31st May 1916\.

**Comment by p5pRT** [comment]
@iabyn - Status changed from 'open' to 'resolved'
